### DISCUSSION TITLE
How to improve typescript intellisense on tRPC 10 ?

### DISCUSSION THREAD

**Discussion by fortezhuo** [trpc/trpc#3024] [Q&A] [answered]
I have file _appRouter.ts on our turborepo apps

```
import { t } from "../util/trpc"
import { userRouter } from "../common/user"
import { captchaRouter } from "./captcha"
import { dominoRouter } from "../common/domino"
import { authRouter } from "../common/auth"
import { mailRouter } from "../common/mail"
import { pmdRouter } from "../common/pmd"
import { workflowRouter } from "../common/workflow"
import { categoryRouter } from "./category"
import { vendor_accountRouter } from "./vendor_account"
import { vendor_approvalRouter } from "./vendor_approval"
import { vendor_infoRouter } from "./vendor_info"

export const appRouter = t.router({
  auth: authRouter,
  pmd: pmdRouter,
  domino: dominoRouter,
  mail: mailRouter,
  workflow: workflowRouter,
  user: userRouter,
  captcha: captchaRouter,
  category: categoryRouter,
  vendor_approval: vendor_approvalRouter,
  vendor_account: vendor_accountRouter,
  vendor_info: vendor_infoRouter,
})

export type AppRouter = typeof appRouter
```
and react.ts file (for client side)

```
import { httpBatchLink, loggerLink } from "@trpc/client"
import { createTRPCNext } from "@trpc/next"
import { NextPageContext } from "next"
import superjson from "superjson"
import type { AppRouter } from "../routers/_app"

function getBaseUrl() {
  if (typeof window !== "undefined") {
    return ""
  }

  return `http://localhost:${process.env.PORT}`
}

export interface SSRContext extends NextPageContext {
  status?: number
}

export const trpc = createTRPCNext<AppRouter, SSRContext>({
  config() {
    return {
      transformer: superjson,
      links: [
        loggerLink({
          enabled: (opts) =>
            process.env.NODE_ENV === "development" ||
            (opts.direction === "down" && opts.result instanceof Error),
        }),

        httpBatchLink({
          url: `${getBaseUrl()}/api/trpc`,
        }),
      ],
    }
  },
  ssr: true,
  responseMeta(opts) {
    const ctx = opts.ctx as SSRContext
    if (ctx.status) {
      return {
        status: ctx.status,
      }
    }
    const error = opts.clientErrors[0]
    if (error) {
      return {
        status: error.data?.httpStatus ?? 500,
      }
    }
    return {}
  },
})
```
And while I try to consume, intellisense takes time to load (all vscode ext have been disabled)
  
![Screen Recording 2022-10-17 at 16 15 48](https://user-images.githubusercontent.com/9128724/196139939-e3a0e131-4d61-4522-9a12-3b1d0e9656a8.gif)

How to improve times of intellisense ?

**Comment by KATT** [comment, marked as answer]
This seems weird to me, we've tested with quote big procedures: https://github.com/trpc/examples-v10-next-big-router

**Reply by fortezhuo** [reply]
I have found complex schema with zod omit, and, etc ... can make typescript intellisense slow

**Reply by fortezhuo** [reply]
I have managed to improve intelisense significantly by ditch prisma-zod-generator, and use forked zod-prisma instead

**Reply by KATT** [reply]
Interesting - what sort of changes did you do to `zod-prisma`?

**Reply by fortezhuo** [reply]
Since prisma-zod-generator generates all schemas based on prisma client's actions, so I replace with forked zod-prisma that only generate based on prisma models. It only generates 1 file each Prisma models. I had to use forked version of mine, because currently zod-prisma seems unmaintained and doesn't support Prisma 4*.

**Reply by KATT** [reply]
Maybe it could also be solved if you used [project references](https://www.typescriptlang.org/docs/handbook/project-references.html) for that generated folder or if you put these schemas in a dedicated package in a monorepo

**Comment by gino** [comment]
What TypeScript version is your VS Code using? I had similar problems without using tRPC and it could be an issue with the language server in some specific TypeScript versions.

**Reply by fortezhuo** [reply]
- Typescript 4.8.4
- VSCode 1.7.2

**Reply by gino** [reply]
I was on the same version and had similar issues when I wasn’t using tRPC. Could you give version 4.7.4 a try to see if this fixes anything?

I have done the same and didn’t notice any issues so far.

**Reply by fortezhuo** [reply]
Intellisense improves better at version 4.7.4. 

But I still need to wait +/- 8s for complex zod schema. I am playing some scenario and found :
- If I remove the complex zod schema and replace them with z.any(), the intellisense improves a lot (I need to wait 1-2 s)
- Using Nx monorepo (integrated repo) will improves DX, compared to Turborepo

**Reply by gino** [reply]
Hmm I see. I think this is not an issue with tRPC since I am experiencing the same issues in a project where I don’t even use tRPC.

There are quite a lot of issues in the official TypeScript repository regarding performance problems but I am not quite sure either.
